File: vnext/Samples/PlaygroundCS/SampleModule.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.ReactNative.Bridge;

namespace Playground
{
  public class SampleModule : NativeModuleBase
  {
    protected override string Name => nameof(SampleModule);

    protected override IReadOnlyDictionary<string, object> Constants { get; }
      = new Dictionary<string, object>()
    {
      { "a", "\"b\"" },
    };

    protected override IReadOnlyList<MethodInfo> Methods { get; }
      = new List<MethodInfo>()
      {
        // Fire and forget Actions
        new MethodInfo(
          "method1", ReturnType.Void,
          (_, __, ___) => {
        Debug.WriteLine("SampleModule.method1()");
        }),
        // Methods with Callbacks
        new MethodInfo(
          "method2", ReturnType.Callback,
          (args, callback, _) => {
            Debug.WriteLine($"SampleModule.method2({String.Join(", ", args)})");

            // Build up a complex object to pass back as arguments when invoking
            // the callback.  They'll automatically be converted to an equivalent
            // type in JavaScript when the JS function is called.
            var parameters = new List<object> {
              84,
              new[] { 0, 1, 2, 3, 4 },
              new[] { 0.02f, 1.0f, 0.42f },
              new[] { "a", "bb" },
              new object[] { new object[] { "First", 0x48 }, "Last" }
            };
            callback(parameters.AsReadOnly());
          }),
        // Referencing a static method
        new MethodInfo("method3", ReturnType.Callback, Method3),
        // Methods with Promise
        new MethodInfo(
          "method4", ReturnType.Promise,
          (args, resolve, reject) => {
            try
            {
              resolve(new object[] {true});
            }
            catch (Exception e)
            {
              reject(new object[] {"0x80000000", e.Message, e.StackTrace, e});
            }
          })
      };

    public static void Method3(
      IReadOnlyList<object> args,
      Callback callback,
      Callback ignored)
    {
      // At the moment the input args are packed into a single JSON-formatted
      // string.
      Debug.WriteLine($"SampleModule.method3({String.Join(", ", args)})");

      // Invoke the callback with some arguments. Using JSON isn't required.  We
      // could have used other basic types instead (e.g. int, double, float,
      // bool, etc.)
      string text = @"{ ""result"" : true }";
      // This uses the built-in Windows.Data.Json.JsonObject, but any
      // object whose ToString() returns a json-formatted text could be used
      // (e.g. the Newtonsoft.Json.Linq.JObject). No automatic conversion
      // is done on a string passed directly as arguments. It comes out as a
      // string on the other side and would need to be parsed (e.g.
      // JSON.parse()).  Simple... what goes in is what comes out.
      var json = Windows.Data.Json.JsonObject.Parse(text);
      callback(new[] { json });
    }

    protected override void Initialize()
    {
      Debug.WriteLine("Initialized");
    }
  }
}

File: packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleModule.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "SampleModule.h"
#include "SampleModule.g.cpp"
#include <winrt\Windows.Data.Json.h>

using namespace winrt;
using namespace Microsoft::ReactNative;
using namespace Microsoft::ReactNative::Bridge;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;

namespace winrt::SampleApp::implementation {

void SampleModule::Initialize() {
  // TODO: Add support on ReactContext to register as an
  // IBackgroundEventListener and/or ILifeCycleEventListener
  OutputDebugStringW(L"Initializing\n");
}

hstring SampleModule::Name() {
  return L"SampleModule";
}

IMapView<hstring, IInspectable> SampleModule::Constants() {
  auto constants = single_threaded_map<hstring, IInspectable>();
  constants.Insert(L"a", box_value(L"\"b\""));
  return constants.GetView();
}

IVectorView<MethodInfo> SampleModule::Methods() {
  auto delegates = single_threaded_vector<MethodInfo>(
      {{L"method1",
        ReturnType::Void,
        [](IVectorView<IInspectable> const & /*ignored*/,
           Callback const & /*ignored*/,
           Callback const & /*ignored*/) {
          // no op
          OutputDebugStringW(L"SampleModule.method1()\n");
        }},
       {L"method2",
        ReturnType::Callback,
        [](IVectorView<IInspectable> const & /*ignored*/,
           Callback const &callback,
           Callback const & /*ignored*/) {
          OutputDebugStringW(L"SampleModule.method2()\n");

          // Build up a complex object to pass back as arguments when invoking
          // the callback.  They'll automatically be converted to an
          // equivalent type in JavaScript when the JS function is called.
          auto integerArray = PropertyValue::CreateInt32Array(
              std::array<int32_t, 5>{0, 1, 2, 3, 4});

          auto floatArray = PropertyValue::CreateSingleArray(
              std::array<float_t, 3>{0.02f, 1.0f, 0.42f});

          auto stringArray = PropertyValue::CreateStringArray({L"a", L"bb"});

          auto mixedArray =
              PropertyValue::CreateInspectableArray(std::array<IInspectable, 2>{
                  box_value(L"First"), box_value(0x48)});

          auto nestedArray = PropertyValue::CreateInspectableArray(
              std::array<IInspectable, 2>{mixedArray, box_value(L"Last")});

          auto callbackArgs =
              single_threaded_vector<IInspectable>({box_value(84),
                                                    integerArray,
                                                    floatArray,
                                                    stringArray,
                                                    nestedArray});

          callback(callbackArgs.GetView());
        }},
       {L"method3", ReturnType::Callback, Method(this, &SampleModule::method3)},
       {L"method4",
        ReturnType::Promise,
        [](IVectorView<IInspectable> const & /*ignored*/,
           Callback const &resolve,
           Callback const &reject /*Promise const &promise*/) {
          //
          try {
            resolve({box_value(true)});
          } catch (std::exception e) {
            hstring what = to_hstring(e.what());
            reject({box_value(L"0x80000000"),
                    box_value(what),
                    box_value(L"stack trace goes here")});
          }
        }}});

  return delegates.GetView();
}

void SampleModule::method3(
    IVectorView<IInspectable> const &args,
    Callback const &callback,
    Callback const & /*ignored*/) {
  if (args.Size() < 1)
    throw hresult_invalid_argument();

  // At the moment the input args are packed into a single JSON-formatted
  // string.
  auto input = args.GetAt(0).as<IPropertyValue>();
  assert(input.Type() == PropertyType::String);
  auto jsonIn = input.GetString();
  OutputDebugStringW(L"SampleModule.method3(");
  OutputDebugStringW(jsonIn.c_str());
  OutputDebugStringW(L")\n");

  // Invoke the callback with some arguments. Using JSON isn't required.  We
  // could have used other basic types instead (e.g. int, double, float, bool,
  // etc.)
  auto text = L"{ \"result\" : true }";
  auto jsonOut = Windows::Data::Json::JsonObject::Parse(text);
  auto callbackArgs = single_threaded_vector<IInspectable>({jsonOut}).GetView();

  callback(callbackArgs);
}

} // namespace winrt::SampleApp::implementation
